### PR TITLE
Fix registration of async functions

### LINKED ISSUE #1191: [Bug]: Async functions are not awaited
### Describe the bug

When defining async functions, the user agent never awaits the async function, it just returns the coroutine. 

### Steps to reproduce

This can be reproduced with the first example at https://github.com/microsoft/autogen/blob/main/notebook/agentchat_function_call_async.ipynb

### Expected Behavior

It should await the async function, not return a coroutine to the group chat. Note that in the logs below it says that the function has started, but it's trivial to add a print function or breakpoint to see that it hasn't (also that it never gives a result).

### Screenshots and logs
```

llm_config = {
    "config_list": config_list,
    'cache_seed': None,
}

coder = autogen.AssistantAgent(
    name="chatbot",
    system_message="For coding tasks, only use the functions you have been provided with. Reply TERMINATE when the task is done.",
    llm_config=llm_config,
)


user_proxy = autogen.UserProxyAgent(
    name="user_proxy",
    system_message="A proxy for the user for executing code.",
    is_termination_msg=lambda x: x.get("content", "") and x.get("content", "").rstrip().endswith("TERMINATE"),
    human_input_mode="NEVER",
    max_consecutive_auto_reply=10,
    code_execution_config={"work_dir": "coding"},
)



@user_proxy.register_for_execution()
@coder.register_for_llm(description="create a timer for N seconds")
async def timer(num_seconds: Annotated[str, "Number of seconds in the timer."]) -> str:
    for i in range(int(num_seconds)):
        time.sleep(1)
        # should print to stdout
    return "Timer is done!"


await user_proxy.a_initiate_chat(  # noqa: F704
    coder,
    message="Create a timer for 5 seconds.",
)
```

------------------------------------

```user_proxy (to chatbot):

Create a timer for 5 seconds.

--------------------------------------------------------------------------------
chatbot (to user_proxy):

***** Suggested tool Call (call_EL6Lt5Qc7H1CtDazCCTYyY0V): timer *****
Arguments: 
{"num_seconds":"5"}
**********************************************************************

--------------------------------------------------------------------------------

>>>>>>>> EXECUTING FUNCTION timer...
user_proxy (to chatbot):

user_proxy (to chatbot):

***** Response from calling tool "timer" *****
<coroutine object timer at 0x00000242110DCAC0>
**********************************************

--------------------------------------------------------------------------------
[c:\Users\10JLW\Anaconda3\lib\site-packages\autogen\agentchat\conversable_agent.py:880](file:///C:/Users/10JLW/Anaconda3/lib/site-packages/autogen/agentchat/conversable_agent.py:880): RuntimeWarning: coroutine 'timer' was never awaited
  _, func_return = self.execute_function(function_call)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
chatbot (to user_proxy):

The timer for 5 seconds has been started.

--------------------------------------------------------------------------------
user_proxy (to chatbot):



--------------------------------------------------------------------------------
chatbot (to user_proxy):

TERMINATE

--------------------------------------------------------------------------------```

### PR DESCRIPTION
**Pull request by davorrunje**
## Why are these changes needed?

Fixes registration of `async` functions.

## Related issue number

Closes #1191 

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

### CODE REVIEW

File: test/agentchat/test_conversable_agent.py [hunk 1]
```diff
@@ -624,6 +634,161 @@ async def exec_sh(script: Annotated[str, "Valid shell script to execute."]):
         assert get_origin(user_proxy_1.function_map) == expected_function_map
 
 
+@pytest.mark.skipif(
+    skip or not sys.version.startswith("3.10"),
+    reason="do not run if openai is not installed or py!=3.10",
+)
+def test_function_restration_e2e_sync() -> None:
```
Comment (davorrunje) [inline test/agentchat/test_conversable_agent.py:641]: added end to end integration test for both `sync` and `async` case

File: test/test_function_utils.py [hunk 2]
```diff
@@ -355,21 +356,41 @@ def test_get_load_param_if_needed_function() -> None:
...
+    assert not asyncio.coroutines.iscoroutinefunction(f)
+
     actual = f(base={"currency": "USD", "amount": 123.45}, quote_currency="EUR")
     assert isinstance(actual[0], Currency)
     assert actual[0].amount == 123.45
     assert actual[0].currency == "USD"
     assert actual[1] == "EUR"
 
 
+@pytest.mark.asyncio
```
Comment (davorrunje) [inline test/test_function_utils.py:376]: added test for `async` wrapper

File: autogen/function_utils.py [hunk 3]
```diff
@@ -319,15 +325,27 @@ def load_basemodels_if_needed(func: Callable[..., Any]) -> Callable[..., Any]:
...
+    def _load_parameters_if_needed(*args: Any, **kwargs: Any) -> Any:
         # load the BaseModels if needed
         for k, f in kwargs_mapping.items():
             kwargs[k] = f(kwargs[k], param_annotations[k])
 
         # call the original function
         return func(*args, **kwargs)
 
-    return load_parameters_if_needed
+    @functools.wraps(func)
```
[the remark this answers is not in the document]
Comment (yenif) [reply autogen/function_utils.py:336]: I've been thinking the break between sync and async reply methods can likely be resolved by using [promisio](https://github.com/miguelgrinberg/promisio) instead as promises can wrap sync and async with the same interface. Haven't had a chance to fully explore it yet.

**Review by yenif — APPROVED** [review]

File: test/agentchat/test_conversable_agent.py [hunk 4]
```diff
@@ -624,6 +634,161 @@ async def exec_sh(script: Annotated[str, "Valid shell script to execute."]):
...
+    timer_mock.assert_called_once_with(num_seconds="2")
+    stopwatch_mock.assert_called_once_with(num_seconds="3")
+
+
+@pytest.mark.skipif(
+    skip or not sys.version.startswith("3.10"),
+    reason="do not run if openai is not installed or py!=3.10",
+)
+@pytest.mark.asyncio()
+async def test_function_restration_e2e_async() -> None:
```
Comment (sonichi) [inline test/agentchat/test_conversable_agent.py:719]:
```suggestion
async def test_function_registration_e2e_async() -> None:
```

File: test/agentchat/test_conversable_agent.py (hunk 1 above)
Comment (sonichi) [reply test/agentchat/test_conversable_agent.py:641]:
```suggestion
def test_function_registration_e2e_sync() -> None:
```

**Pull request merged**